### ISSUE TITLE
Number of wakeup notifications using rndv is ~5 but ~2 for eager

### ISSUE THREAD

**Issue by Akshay-Venkatesh** [openucx/ucx#3542]
Hi,

We're trying to reduce progress overhead using ucxpy. As we use wakeup mechanism to check for request completion we want to cut down the number of times we wake up per request if possible. We notice that when the request concerns eager messages the number of times wakeup notifications is observed is upper bounded by 2. When the request concerns rendezvous messages, the number of wakeup notifications is upper bounded by 5. This may be expected given that the latter has control message exchange in addition to data payload and this results in additional completion events and subsequently more notifications. 

I wanted to check if it is possible to have a single notification per request completion. Is it possible to ignore some of the network events that are passed down to ucp_worker file descriptor in order to achieve this?  I don't know if there exists a way to achieve this because I would think that wakeup on control message is required to trigger application to call progress and subsequently complete some action required based on the control message before the actual ucp request can be completed. What are your thoughts on this? Is there an alternative way where I can cut down the cost of checking if request is completed when operating in wakeup mode?

**Comment by shamisp** [comment]
I'm surprised that with eager we get two.
Is this something that you get directly from file descriptor ?

**Comment by Akshay-Venkatesh** [comment]
@shamisp I think I'm wrong about the eager number. I modified src/tools/perf/lib/ucp_tests.cc to count number of wakeup instances. I think by running tag_lat, I'm wrongly counting the wakeup instances of both the sender and receiver. That would best explain why eager count is 2 (one at send and one at receiver), and rndv-get count is 5 (2 at sender and 3 at receiver). My question still stands if it's at all possible to cut down the number of notifications to 1 per request.

**Comment by yosefe** [comment]
@Akshay-Venkatesh maybe in theory it can be possible to reduce it to 1 on sender (RNDV_ACK) and 2 on the receiver (RNDV_RTS and RDMA_READ send completion).

**Comment by Akshay-Venkatesh** [comment]
> @Akshay-Venkatesh maybe in theory it can be possible to reduce it to 1 on sender (RNDV_ACK rx) and 2 on the receiver (RNDV_RTS rx and RDMA_READ tx completion).

This would still be very useful for our case and wherever wakeup mode used.

**Comment by Akshay-Venkatesh** [comment]
Another thought - Does hardware tag matching help in this regard?

**Comment by yosefe** [comment]
HW TM would not help that, since still need 2 events on RX side: notify when tag is matched to remove it from SW list, and also when data has arrived.

**Comment by Akshay-Venkatesh** [comment]
Makes sense. Seems like 1 wakeup notification per request wouldn't be possible. If yes, I'll close this issue for now.

**Comment by yosefe** [comment]
i don't see how 1 wakeup for sender side rndv could be possible, but we still can keep this open with the goal to make sure the rndv wakeup count is minimal

**Comment by shamisp** [comment]
I was wondering if there is a way to filter this through WID field, and just mask some of the completions  from exposure.

**Comment by yosefe** [comment]
WID is SW-level concept.. and AFAIK there is no way to filter RX completions

**Comment by yosefe** [comment]
Will check if solicited events with HWTM can reduce RX wakeup count
